README
* INTRODUCTION

  Rudel is a collaborative editing environment for GNU Emacs. Its
  purpose is to share buffers with other users in order to edit the
  contents of those buffers collaboratively. Rudel supports multiple
  backends to enable communication with other collaborative editors
  using different protocols, though currently Obby (for use with the
  Gobby editor) is the only fully-functional one.

  Since Rudel is not an application, but an extension to Emacs, it is
  not started and used like most applications (not even Emacs
  applications like Gnus). Rudel mostly works in the background to
  change the behavior of the set of Emacs buffers for which it has
  been activated.

  The user interface consists of a set of key bindings, a menu entry
  and some visual status indicators, which are added to the text,
  header line and/or mode line of buffers for which Rudel has been
  activated.

* GETTING STARTED

  Assuming Rudel has already been installed (see file:INSTALL) and
  auto loading has been set up, a global Rudel mode can be enabled as
  follows:

  : M-x global-rudel-minor-mode

  This will enable Rudel's key bindings and add a menu entry. To
  enable this mode permanently, the following fragment can be added to
  the Emacs initialization file (usually file:~/.emacs):

  #+BEGIN_SRC emacs-lisp
  (global-rudel-minor-mode 1)
  #+END_SRC

** JOINING A SESSION

   : M-x rudel-join-session [ C-c c j ]

   Depending on the installed Rudel backends, system environment and
   configuration, a number of questions will be asked, followed by an
   attempt to join session described by your answers.

   A typical example of the questions asked when joining a session may
   look like this:

   #+BEGIN_EXAMPLE
   Server: localhost RET
   Port (default 6522): RET
   Username: jan RET
   Color: light sky blue RET
   Use Encryption (y or n): n RET
   Global Password: RET
   User Password: RET
   #+END_EXAMPLE

   *IMPORTANT*: For sessions using the obby backend (like in the
   example above), the following restriction has to be taken into
   account:
   + When the server is Rudel inside an Emacs process:
     Encryption cannot be used currently in this case. Consequently
     the answer to the `Use Encryption (y or n):' prompt above has to
     be `n RET'.
   + When the server is a Gobby process:
     Gobby only supports encrypted connections. So the answer has to
     be `y RET' is this case.

   It is possible to configure frequently used sessions using the
   customization options `rudel-configured-sessions'. When one or more
   sessions are configured, `rudel-join-session' will provide choices
   like "my-configured-session", ... and "ask-protocol". Selecting
   "ask-protocol" invokes the behavior described above. Selecting one
   of the configured sessions connects to that session without asking
   for all the data.

** HOSTING A SESSION

   : M-x rudel-host-session [ C-c c h ]

   Note that the session starts out without any participating users
   (This is sometimes referred to as being a dedicated server). If you
   want to participate in the session you host, you have to join it as
   described above.

* REPORTING BUGS AND GETTING HELP

** EmacsWiki

   The EmacsWiki page about Rudel has
   [[http://www.emacswiki.org/emacs/Rudel#toc8][a section for feedback and questions]].
   Feel free leave feedback or ask questions there. If a reply does
   not occur promptly, try writing to the mailing list of visiting the
   IRC channel.

** Issue Tracker

   Bugs can be reported using the issue tracker on the sourceforge
   project page: http://sourceforge.net/tracker/?group_id=249139

** Mailing List

   Another possibility for getting help and reporting problems is
   writing to the Rudel mailing list:
   http://sourceforge.net/mail/?group_id=249139

** IRC Channel

   Rudel users and developers can also be reached on the #rudel IRC
   channel on the freenode network: irc:/irc.freenode.net:6667/#rudel

* KNOWN ISSUES

  + Publishing eshell buffers will cause your session to be
    disconnected since eshell disables the hooks that Rudel uses to
    catch changes to the buffer. As a workaround, you can use M-x
    ansi-term or another terminal emulator.

* LICENSE

  This file is part of Rudel.

  Rudel is free software: you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Rudel is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
  License for more details.

  You should have received a copy of the GNU General Public License
  along with Rudel.  If not, see <http://www.gnu.org/licenses/>.

#+TITLE:   Rudel README
#+AUTHOR:  Jan Moringen
#+STARTUP: showeverything

# Local variables:
# mode: org
# end:


ChangeLog
2010-07-25  Joseph Corneli  <holtzermann17@gmail.com>

	Renamed rudel-end-session to rudel-leave-session
	* rudel.el (rudel-end-session): changed
	rudel-end-session to rudel-leave-session
	* rudel-mode.el (rudel-minor-keymap): likewise

	Renamed rudel-unpublish-buffer to rudel-unsubscribe
	* rudel.el (rudel-attach-to-buffer): changed
	rudel-unpublish-buffer to rudel-unsubscribe
	(rudel-detach-from-buffer): likewise
	(rudel-unsubscribe): likewise
	* rudel-mode.el (rudel-minor-keymap): likewise
	* rudel-interactive.el (rudel-allocate-buffer-clear-existing):
	likewise

        Updates to INSTALL file.
	* INSTALL: Rearranged and updated instructions so they make more
	sense.

2010-07-02  Jan Moringen  <jmoringe@techfak.uni-bielefeld.de>

	Proper comments for local variables in README, INSTALL
	* README: added proper comments to local variables
	* INSTALL: added proper comments to local variables

2010-06-15  Jan Moringen  <scymtym@users.sourceforge.net>

	Require rudel-util.el in rudel-mode.el
	* rudel-mode.el (header): minor fix in history section
	(require rudel-util): new require; for object-{add,remove}-hook

	New milestones and tags, reassigned items in TODO
	* TODO (Future): reassigned and updated some items
	(Milestone rudel-0.5): new milestone
	(Milestone rudel-0.4): added tag 0.4; scheduled tasks "show cursor
	positions of other users", "integrate unit tests", "only read
	color hue"; minor fixes
	(Milestone rudel-0.3): added tag 0.3
	(Milestone rudel-0.2): added tags 0.2 and released
	(Milestone rudel-0.1): added tags 0.1 and released

	Hopefully fixed some notorious problems in rudel-compile.el
	* rudel-compile.el (require cl): new require; for flet
	(toplevel compilation): let-bind filename since there have been
	some compilation problems related to `filename' being void; inject
	requires for cl and rudel-backend into rudel-loaddefs.el

	Mention Debian/Ubuntu packages in INSTALL
	* INSTALL (REQUIREMENTS): mention Debian/Ubuntu packages for
	gnutls-cli and Avahi
	(footer): added settings startup, title and author

	Removed redundant checks in -factory::rudel-suitable-backends
	* rudel-backend.el
	(rudel-backend-factory::rudel-suitable-backends): do not check the
	cdr being an object twice

	Better handling of loading errors in rudel-backend.el
	* rudel-backend.el (header): minor fix in history
	(rudel-backend-factory::rudel-all-backends): added optional
	argument only-loaded; only return unloaded backends if requested
	(rudel-backend-factory::rudel-suitable-backends): only consider
	loaded backends
	(rudel-backend-factory::rudel-load-backends): if the loading
	fails, install the error message as a symbol property of the
	backend name symbol
	(rudel-backend-dump): format backends differently depending on
	whether they have been loaded successfully or not
	(rudel-backend--format-backend-normal): new function; format a
	loaded or unloaded backend
	(rudel-backend--format-backend-error): new function; format a
	backend for which loading has failed

2010-06-14  Jan Moringen  <scymtym@users.sourceforge.net>

	Added some docstrings in infinote/rudel-infinote-group.el
	* infinote/rudel-infinote-group.el
	(rudel-infinote-group::connection): added documentation string
	(rudel-infinote-group::rudel-register-state): added documentation
	string
	(rudel-infinote-group::rudel-send): added documentation string

	Improved session startup in infinote/rudel-infinote{,client}.el
	* infinote/rudel-infinote.el
	(rudel-infinote-backend::rudel-connect): use `rudel-wait' to
	ensure the connection is ready for use; handle missing :host
	property
	(rudel-infinote-backend::rudel-make-node): do not set the :group
	slot when creating nodes
	* infinote/rudel-infinote-client.el
	(rudel-infinote-client-connection::initialize-instance): cosmetic
	changes
	(rudel-infinote-client-connection::rudel-make-and-add-group):
	added a comment
	(rudel-infinote-client-connection::rudel-make-and-add-node): added
	a comment
	(rudel-infinote-client-connection::rudel-receive): added comments
	(rudel-infinote-client-connection::rudel-wait): new method; block
	until the connection is ready for use; this is determined by
	monitoring the state of the directory group

2010-06-13  Jan Moringen  <scymtym@users.sourceforge.net>

	Improved docstring of rudel-xmpp-backend::rudel-make-connection
	* xmpp/rudel-xmpp.el (rudel-xmpp-backend::rudel-make-connection):
	improved documentation string

	Removed temporary slots id and status from rudel-infinote-user
	* infinote/rudel-infinote-user.el (rudel-infinote-user::id):
	removed; was temporary
	(rudel-infinote-user::status): removed; was temporary

	Added new class rudel-hook-state-machine to rudel-state-machine.el
	* rudel-state-machine.el (header): fixed keyword; added version
	0.2 to history section
	(require rudel-util): new require; used by the hook state machine
	(rudel-state-machine::state): added writer method
	`rudel-set-state'
	(rudel-state-machine::initialize-instance): use `rudel-set-state'
	(rudel-state-machine::rudel-switch): use `rudel-set-state'
	(rudel-hook-state-machine): new class; a state machine that runs
	hooks when it accepts input or when it switches states

2010-06-12  Jan Moringen  <scymtym@users.sourceforge.net>

	Added docs, comments to infinote/rudel-infinote-group-document.el
	* infinote/rudel-infinote-group-document.el
	(rudel-infinote-group-document-state-idle::rudel-infinote/sync-begin):
	added documentation string; added some comments
	(rudel-infinote-group-document-state-idle::rudel-infinote/session-close):
	added documentation string; added some comments
	(rudel-infinote-group-document-state-synchronizing::rudel-infinote/sync-user):
	cleanup; updated some comments
	(rudel-infinote-group-document-state-synchronizing::rudel-infinote/sync-segment):
	added documentation string; added some comments
	(rudel-infinote-group-document-state-synchronizing::rudel-infinote/sync-end):
	added some comments

rudel-backend.el
;;; rudel-backend.el --- A generic backend management mechanism for Rudel
;;
;; Copyright (C) 2009, 2010 Jan Moringen
;;
;; Author: Jan Moringen <scymtym@users.sourceforge.net>
;; Keywords: Rudel, backend, factory
;; X-RCS: $Id:$
;;
;; This file is part of Rudel.
;;
;; Rudel is free software: you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; Rudel is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with Rudel. If not, see <http://www.gnu.org/licenses>.


;;; Commentary:
;;
;; This file contains a generic mechanism that handles registration,
;; query and instantiation of Rudel backends for any number of
;; functional categories.
;;
;; The class and collaboration design is as follows: for each
;; category, which is identified by a symbol, there is a factory
;; object (an instance of `rudel-backend-factory') that is responsible
;; for creating backend objects of the category.  Examples of
;; categories are 'transport', 'protocol' and 'session-initiation'.
;; In addition to creating backend object, factories also allow
;; querying backends based on desired capabilities and load backend
;; implementations only when required.


;;; History:
;;
;; 0.1 - Initial version


;;; Code:
;;

(eval-when-compile
  (require 'cl))

(require 'warnings)

(require 'eieio)


;;; Class rudel-backend
;;

(defclass rudel-backend ()
  ((version      :initarg  :version
		 :type     list
		 :documentation
		 "A list of the form (MAJOR MINOR [MICRO
WHATEVER*]) describing the version of the backend.")
   (capabilities :initarg  :capabilities
		 :type     list
		 :initform nil
		 :documentation
		 "A list of symbols, or lists whose car is a
symbol, that each describe one capability of the backend."))
  "Base class for backend classes."
  :abstract t)

(defmethod rudel-capable-of-p ((this rudel-backend) capability)
  "Return t if the backend THIS is capable of CAPABILITY."
  (with-slots (capabilities) this
    (member capability capabilities)))


;;; Class rudel-backend-factory
;;

;;;###autoload
(defclass rudel-backend-factory ()
  ((backends  :initarg   :backends
	      :type       hash-table
	      :documentation
	      "Mapping of symbolic names to classes (prior to
instantiation) or objects (after instantiation) for all backends
known to the factory object.")
   (factories :type       hash-table
	      :allocation :class
	      :documentation
	      "Mapping of backend categories to responsible
factory objects."))
  "Factory class that holds an object for each known backend
category. Objects manage backend implementation for one backend
category each.")
(oset-default rudel-backend-factory factories
	      (make-hash-table :test #'eq))

(defmethod initialize-instance ((this rudel-backend-factory) &rest slots)
  "Initialize slots of THIS with SLOTS."
  (when (next-method-p)
    (call-next-method))
  (oset this :backends (make-hash-table :test #'eq)))

;;;###autoload
(defmethod rudel-get-factory :static ((this rudel-backend-factory)
				      category)
  "Return the factory responsible for CATEGORY.
If there is no responsible factory, create one and return it."
  (with-slots (factories) this
    (or (gethash category factories)
	(puthash category (rudel-backend-factory category) factories)))
  )

;;;###autoload
(defmethod rudel-add-backend ((this rudel-backend-factory)
			      name class &optional replace)
  "Add factory class CLASS with name NAME to THIS.
if REPLACE is non-nil, replace a registered implementation of the
same name."
  (with-slots (backends) this
    (when (or (not (gethash name backends))
	      replace)
      (puthash name class backends))))

(defmethod rudel-get-backend ((this rudel-backend-factory) name)
  "Return backend object for name NAME or nil if there is none.
The returned backend is of the form (NAME . OBJECT).

Backends are loaded, if necessary."
  ;; Load all available backends
  (rudel-load-backends this)

  ;; Find the backend and return it.
  (with-slots (backends) this
    (let ((backend (gethash name backends)))
      (when backend
	(cons name backend))))
  )

(defmethod rudel-all-backends ((this rudel-backend-factory)
			       &optional only-loaded)
  "Return a list of all backends registered with THIS.
Each list element is of the form (NAME . CLASS-OR-OBJECT).
If optional argument ONLY-LOADED is non-nil, return only elements
for which CLASS-OR-OBJECT is an object."
  (let ((backend-list))
    (with-slots (backends) this
      (maphash (lambda (name class)
		 (when (or (not only-loaded)
			   (object-p class))
		   (push (cons name class) backend-list)))
	       backends))
    backend-list)
  )

(defmethod rudel-suitable-backends ((this rudel-backend-factory) predicate)
  "Return a list of backends which satisfy PREDICATE.
Each list element is of the form (NAME . OBJECT).
Backends are loaded, if necessary."
  ;; Load all available backends.
  (rudel-load-backends this)

  ;; Retrieve and return all backends, filtering the list using
  ;; PREDICATE. Backends that could not be loaded, are ignored.
  (if predicate
      (remove-if-not
       (lambda (cell)
	 (funcall predicate (cdr cell)))
       (rudel-all-backends this t))
    (rudel-all-backends this t))
  )

(defmethod rudel-load-backends ((this rudel-backend-factory))
  "Load backends in THIS factory if necessary.
Loading errors are not reported explicitly, but can be detected
by checking for backends that still are classes rather than
objects."
  ;; Map lambda that loads unloaded backends over all backends. Store
  ;; objects back after loading.
  (with-slots (backends) this
    (maphash
     (lambda (name class)
       (unless (object-p class)
	 (condition-case error
	     (puthash name (make-instance
			    class (symbol-name name)) backends)
	   (error
	    ;; Store this error on the name symbol of the backend for
	    ;; later display.
	    (put name 'rudel-backend-last-load-error
		 (error-message-string error))

	    ;; Emit a warning.
	    (display-warning
	     '(rudel backend)
	     (format "Could not load backend `%s': %s"
		     name
		     (error-message-string error))
	     :warning)))))
	 backends))
  )


;;; High-level frontend functions
;;

(defsubst rudel-backend-cons-p (cell)
  "Check whether CELL is a cons of a backend name and object."
  (and (consp cell)
       (symbolp (car cell))
       (object-p (cdr cell))))

;;;###autoload
(defun rudel-backend-get (category name)
  "A shortcut for getting backend NAME of category CATEGORY.
The returned backend is of the form (NAME . OBJECT)."
  (rudel-get-backend (rudel-backend-get-factory category) name))

;;;###autoload
(defun rudel-backend-get-factory (category)
  "A shortcut for getting the factory object for CATEGORY."
  (rudel-get-factory rudel-backend-factory category))

(defun rudel-backend-suitable-backends (category predicate)
  "Return backends from category CATEGORY that satisfy PREDICATE.
Each list element is of the form (NAME . OBJECT)."
  (rudel-suitable-backends
   (rudel-backend-get-factory category)
   predicate))

(defun rudel-backend-choose (category &optional predicate)
  "Choose a backend from CATEGORY satisfying PREDICATE automatically or by asking the user.
The returned backend is of the form (NAME . CLASS-OR-OBJECT)."
  (let ((backends (rudel-backend-suitable-backends
		   category predicate)))
    (unless backends
      (error "No backends available"))

    (if (= (length backends) 1)
	;; If there is only one backend, we can choose that one right
	;; away displaying a message to avoid confusing the user.
	(let ((backend (nth 0 backends)))
	  (message "Using backend `%s'" (symbol-name (car backend)))
	  (sit-for 0.5)
	  backend)

      ;; When we have more than one backend, we have to ask the user,
      ;; which one she wants.
      (require 'rudel-interactive)
      (rudel-read-backend
       backends
       (format "%s backend: " (capitalize (symbol-name category)) )
       'object)))
  )


;;; User interaction functions
;;

(defun rudel-backend-dump (&optional load)
  "Create display information about backends in a buffer.
If LOAD is non-nil, load all backends before display. This makes
available information available for the backends"
  (interactive "p")
  ;; Setup a new buffer.
  (set-buffer (get-buffer-create "*Rudel Backends*"))
  (erase-buffer)
  (set-window-buffer nil (current-buffer))
  (maphash
   (lambda (category factory)
     ;; Load backends if requested.
     (unless (zerop load)
       (rudel-load-backends factory))

     ;; Insert header for this category.
     (insert (propertize
	      (format "Category %s\n" category)
	      'face 'bold))
     (insert (apply #'format
		    "  %-20s %-6s %-7s %s\n"
		    (mapcar
		     (lambda (header)
		       (propertize header 'face 'italic))
		     '("name" "loaded" "version" "capabilities"))))

     ;; Insert all backends provided by this factory.
     (dolist (backend (rudel-all-backends factory))
       (if (or (object-p (cdr backend))
	       (null (get (car backend)
			  'rudel-backend-last-load-error)))
	   (insert (rudel-backend--format-backend-normal backend))
	 (insert (rudel-backend--format-backend-error backend))))

     ;; One empty line between backend categories.
     (insert "\n"))
   (oref rudel-backend-factory factories))
  )

(defun rudel-backend--format-backend-normal (backend)
  "Format name, version and other properties of BACKEND."
  (format "  %-20s %-6s %-7s (%s)\n"
	  ;; Backend name
	  (propertize
	   (symbol-name (car backend))
	   'face 'font-lock-type-face)
	  ;; Backend loading status
	  (propertize
	   (prin1-to-string (object-p (cdr backend)))
	   'face 'font-lock-variable-name-face)
	  ;; Backend version
	  (propertize
	   (if (object-p (cdr backend))
	       (mapconcat #'prin1-to-string
			  (oref (cdr backend) :version)
			  ".")
	     "?")
	   'face 'font-lock-constant-face)
	  ;; Backend capabilities
	  (propertize
	   (if (object-p (cdr backend))
	       (mapconcat #'prin1-to-string
			  (oref (cdr backend) :capabilities)
			  " ")
	     "?")
	   'face 'font-lock-constant-face))
  )

(defun rudel-backend--format-backend-error (backend)
  "Format name and loading error message of BACKEND."
  (format "  %-20s %s %s\n"
	  ;; Backend name
	  (propertize
	   (symbol-name (car backend))
	   'face 'font-lock-type-face)
	  ;; Loading error message
	  (propertize
	   "error:"
	   'face 'font-lock-warning-face)
	  (propertize
	   (get (car backend) 'rudel-backend-last-load-error)
	   'face 'font-lock-comment-face))
  )

(provide 'rudel-backend)
;;; rudel-backend.el ends here
